### ISSUE TITLE
Issue #38: Distribution Cut Prediction Model v2.0 - Enhanced Features and Macro Integration

### ISSUE THREAD

**Issue by reggiechan74** [reggiechan74/issuer-credit-analysis#38]
## Overview

Enhance the distribution cut prediction model (currently 100% accuracy on n=9 dataset) with advanced cash flow metrics, macro variables, market risk signals, and expanded training data for production deployment at scale.

**Current Model (v1.0):**
- ✅ 100% accuracy (9/9 correct predictions)
- ✅ F1-Score: 1.000 (perfect classification)
- ✅ Feature set: AFFO payout, interest coverage, debt/assets, debt/EBITDA, occupancy, sector
- ✅ Training dataset: n=9 REITs (6 cuts, 3 controls)
- ⚠️ Limitation: Lacks cash flow metrics (ACFO, AFCF, burn rate), macro context, market signals

**Version 2.0 Goals:**
- Expand training dataset: n=9 → n=28 REITs ✨ **Updated (Issue #39 enables 28 vs original target 20-30)**
- Add cash flow features: ACFO, AFCF, burn rate, cash runway
- Integrate macro variables: interest rates, credit stress scoring, economic cycle ✅ **Automated (Issue #39)**
- Add market risk signals: price stress, volatility, momentum ✨ **NEW (Issue #39)**
- Add recovery patterns: time to recover, recovery level % ✨ **NEW (Issue #39)**
- Add debt maturity risk: time-to-maturity, refinancing cliff analysis

**Timeline Revised:** ~~10 weeks~~ → **5-6 weeks (40-50% reduction via Issue #39 automation)**

---

### 🚀 Dataset Expansion Automated (95% Time Savings)

**Before (Original Issue #38 Plan):**
- Week 1-2: Manually research and extract 15-20 distribution cuts
- Estimated time: 3 weeks (120+ hours)

**After (Issue #39 Automation):**
- `build_distribution_cut_dataset.py` - Fully automated
- Actual time: 2-3 hours
- **Result:** 28 REITs processable (exceeds target)

**Test Results:**
- 10-REIT validation: 9/10 successful (90%)
- 9 distribution cuts detected across 6 REITs
- 10-15 years of dividend history per REIT

### 🚀 Macro Data Integration Automated (100% Time Savings)

**Before (Original Issue #38 Plan):**
- Week 4-5: Manually collect Bank of Canada rates, merge features
- Estimated time: 2 weeks

**After (Issue #39 Automation):**
- `openbb_macro_monitor.py` - Fully automated
- Features delivered:
  - Canada policy rate (Bank of Canada Valet API - FREE)
  - US Federal Reserve rate (FRED API - FREE)
  - Rate change 12-month
  - Rate cycle classification (TIGHTENING, EASING, STABLE)
  - Credit stress score (0-100 scale)
  - Canada-US rate differential

**Test Results:**
- Current: 2.75% Canada (EASING, -175 bps)
- Current: 4.22% US (MILD EASING, -91 bps)
- Spread: -147 bps (typical range)

### ✨ NEW: Market Risk Signals (Not Originally Planned)

**Delivered by Issue #39:**
- `openbb_market_monitor.py` - Market risk analysis
- Features available:
  - Price stress detection (% decline from 52-week high)
  - Risk scoring (0-100 composite scale)
  - Volatility (30/90/252-day windows)
  - Momentum (3/6/12-month returns)
  - Volume trend analysis

**Hypothesis:** Market signals may provide **early warning** 1-2 quarters before fundamentals deteriorate.

**Test Results:**
- Slate Office: 65/100 (HIGH risk, 40.6% decline)
- RioCan REIT: 5/100 (VERY LOW risk, 5.4% decline)

### ✨ NEW: Recovery Pattern Analysis (Not Originally Planned)

**Delivered by Issue #39:**
- Recovery timeline tracking (months to restore)
- Recovery level percentage (current vs pre-cut)
- Recovery status classification
- Management discipline signals

**Hypothesis:** Recovery pace may differentiate conservative vs aggressive management, predict repeat cut risk.

**Test Results:**
- RioCan: 80.4% recovery from 2021 COVID cut (conservative)
- Artis: 2018 cut fully recovered in 49 months, 2023 cut 47.6% recovery

---

### Phase 1: Automated Dataset Creation (1 week) ✅ **2 weeks saved**

**Week 1: Dataset Generation**
- [x] Run `build_distribution_cut_dataset.py` on 28 REITs → ~2-3 hours ✅ **Automated (Issue #39)**
- [x] Generate market risk data for all REITs → ~2 hours ✅ **Automated (Issue #39)**
- [x] Generate macro data for observation dates → ~1 hour ✅ **Automated (Issue #39)**
- [ ] Manual: Collect GDP, unemployment, CBRE cap rates → 2-3 days
- [ ] Merge all features into training_dataset_v2.csv

**Deliverable:** training_dataset_v2.csv with distribution cuts, market risk, and macro features

### Phase 2: Cash Flow Feature Engineering (1 week)

**Week 2: ACFO/AFCF/Burn Rate Calculation**
- [ ] Extract detailed cash flow data (CFI, CFF components) for all 28 REITs
- [ ] Calculate ACFO for all observations (using Phase 3 pipeline)
- [ ] Calculate AFCF for all observations
- [ ] Calculate burn rate and cash runway
- [ ] Add cash flow features to training_dataset_v2.csv

**Deliverable:** Complete feature set (36 features) ready for modeling

### Phase 3: Feature Selection & Model Training (2 weeks)

**Week 3: Baseline Retraining**
- [ ] Train v1.0 feature set (6 features) on expanded dataset (n=28)
- [ ] Evaluate performance degradation (expect 85-90% vs 100% - acceptable)
- [ ] Establish baseline for comparison
- [ ] SHAP analysis on v1.0 features

**Week 4: Enhanced Feature Testing**
- [ ] Test cash flow features alone (ACFO, AFCF, burn rate)
- [ ] Test market risk features alone (price stress, volatility, momentum)
- [ ] Test enhanced macro features (rates + credit stress score)
- [ ] Test recovery pattern features (recovery time, level %)
- [ ] Test combined feature set (36 features)
- [ ] SHAP analysis: Feature importance ranking for all feature sets

**Deliverable:** Feature importance analysis, optimal feature set identified

### Phase 4: Validation & Deployment (1-2 weeks)

**Week 5: Model Validation**
- [ ] Temporal cross-validation (train 2015-2020, test 2021-2024)
- [ ] Sector-specific validation (office vs retail vs residential)
- [ ] Macro regime testing (low rate 2015-2021 vs high rate 2022-2024)
- [ ] Hyperparameter tuning with larger dataset (n=28)

**Week 6: Production Deployment (Optional)**
- [ ] Update prediction interface (predict_distribution_cut_v2.py)
- [ ] Document new features and usage
- [ ] Create model refresh pipeline (automated macro/market data updates)
- [ ] Deployment guide and API documentation

**Deliverable:** Production-ready v2.0 model with comprehensive documentation

---

### v1.0 Features (6 features) ✅
```python
affo_payout_ratio        # AFFO / Distributions
interest_coverage        # NOI / Interest Expense
debt_to_assets          # Total Debt / Total Assets
debt_to_ebitda          # Net Debt / EBITDA
occupancy_rate          # % occupied
sector                  # Retail, Office, Industrial, Residential, Diversified
```

### v2.0 New Features - Cash Flow (9 features)
```python
acfo_per_unit           # Adjusted Cash Flow from Operations (per unit)
afcf_per_unit           # Adjusted Free Cash Flow (per unit)
afcf_payout_ratio       # AFCF / Distributions (%)
afcf_self_funding_ratio # AFCF / (Debt Service + Distributions)
acfo_margin             # ACFO / Revenue (%)
monthly_burn_rate       # Cash depletion rate (when AFCF < obligations)
cash_runway_months      # Available cash / Monthly burn rate
burn_rate_applicable    # Boolean: 1 if burning cash, 0 if self-funding
liquidity_risk_score    # 1-4 scale (LOW, MODERATE, HIGH, CRITICAL)
```

### v2.0 New Features - Macro (7 features) ✅ **Automated (Issue #39)**
```python
current_policy_rate     # ✅ Bank of Canada overnight rate
rate_change_12m         # ✅ Change in policy rate over 12 months (bps)
rate_cycle              # ✅ TIGHTENING, EASING, STABLE (encoded)
credit_stress_score     # ✅ 0-100 scale (macro credit environment)
ca_us_rate_spread       # ✅ Canada - US rate differential (bps)
gdp_growth_yoy          # ⚠️ Manual: Canada real GDP growth YoY (%)
unemployment_rate       # ⚠️ Manual: Canada unemployment rate (%)
```

### v2.0 New Features - Market Risk (5 features) ✨ **NEW (Issue #39)**
```python
price_stress_pct        # ✅ % decline from 52-week high
market_risk_score       # ✅ 0-100 composite risk score
volatility_90d          # ✅ Annualized 90-day volatility (%)
momentum_3m             # ✅ 3-month price return (%)
volume_trend            # ✅ Trading volume change vs 90-day average
```

### v2.0 New Features - Recovery Patterns (4 features) ✨ **NEW (Issue #39)**
```python
recovery_months         # ✅ Months to restore distribution post-cut
recovery_level_pct      # ✅ Current distribution as % of pre-cut level
prior_cut_count         # ✅ Number of cuts in prior 5 years
recovery_status         # ✅ Full, Partial 80%+, Partial 50-80%, Minimal (encoded)
```

### v2.0 New Features - Debt Maturity (5 features)
```python
debt_maturity_12m       # ⚠️ Debt maturing within 12 months (% of total)
debt_maturity_24m       # ⚠️ Debt maturing within 24 months (% of total)
weighted_avg_maturity   # ⚠️ Weighted average years to maturity
debt_to_assets_cushion  # ⚠️ Covenant limit - Actual ratio (bps)
covenant_breach_risk    # ⚠️ Boolean: 1 if within 5% of any covenant
```

**Total v2.0 Feature Count: 36 features (6x expansion from v1.0)**

**Automation Status:**
- ✅ **Fully Automated (Issue #39):** 16 features (distribution cuts, market risk, core macro)
- ⚠️ **Partially Automated (Pipeline):** 9 features (cash flow - Phase 1-3 pipeline)
- ⚠️ **Manual Collection:** 11 features (GDP, unemployment, cap rates, debt maturity, covenants)

---

### Model Performance (on n=28 dataset)
- ✅ Accuracy ≥ 85% (acceptable given larger, more diverse dataset)
- ✅ Precision ≥ 90% (minimize false positives - no unnecessary alarms)
- ✅ Recall ≥ 80% (catch most cuts before they happen)
- ✅ F1-Score ≥ 0.85

**Note:** Some degradation from v1.0's perfect 100% is expected and acceptable - model becomes more generalizable.

### Feature Insights
- ✅ Cash flow features (ACFO/AFCF/burn rate) have SHAP importance > 0.5
- ✅ Macro features improve F1-score vs v1.0 baseline
- ✅ **NEW:** Market risk features provide early warning (validate hypothesis)
- ✅ **NEW:** Recovery patterns predict repeat cut risk (validate hypothesis)
- ✅ Debt maturity/covenant features add incremental value

### Production Readiness
- ✅ Automated macro data refresh pipeline (Issue #39 tools)
- ✅ Automated market risk monitoring (Issue #39 tools)
- ✅ Prediction interface with SHAP explanations
- ✅ Comprehensive documentation
- ✅ Backtesting validates temporal stability

---

## Timeline Comparison

| Phase | Original Plan | Revised Plan | Time Saved |
|-------|---------------|--------------|------------|
| Dataset Expansion | 3 weeks | 1 week | **2 weeks** (Issue #39 automation) |
| Macro Integration | 2 weeks | 0 weeks | **2 weeks** (Issue #39 automation) |
| Model Development | 3 weeks | 2 weeks | 1 week (smaller dataset, faster iteration) |
| Validation | 2 weeks | 1-2 weeks | 0-1 week |
| **Total** | **10 weeks** | **5-6 weeks** | **4-5 weeks (40-50% reduction)** |

---

### Resolved by Issue #39 ✅
- ✅ Distribution cut database → Automated via `build_distribution_cut_dataset.py`
- ✅ Bank of Canada rate data → Automated via `openbb_macro_monitor.py`
- ✅ US Federal Reserve rate data → Automated via `openbb_macro_monitor.py`
- ✅ Market risk data → Automated via `openbb_market_monitor.py`
- ✅ Recovery pattern analysis → Automated via `openbb_data_collector.py`

### Still Required ⚠️
- ⚠️ Statistics Canada (GDP, unemployment) - Manual API calls or CSV download
- ⚠️ CBRE cap rate data - Manual collection from quarterly market reports
- ⚠️ Debt maturity schedules - Phase 1-3 extraction from MD&A disclosures
- ⚠️ Covenant data - Phase 1-3 extraction from credit agreements/MD&A

### Technical (Unchanged)
- Phase 1-3 pipeline (PDF → JSON → Metrics) for cash flow extraction
- LightGBM v4.0+ (gradient boosting)
- SHAP v0.42+ (feature importance and explainability)
- Pandas, NumPy, Scikit-learn

---

### Risk 1: Market Risk Features May Not Be Predictive
**Risk:** Market signals (price stress, volatility) may lag fundamentals, not lead
**Probability:** MEDIUM
**Mitigation:** Test hypothesis explicitly - train model with/without market features, compare F1-scores
**Fallback:** Exclude market features if SHAP importance < 0.3

### Risk 2: Larger Dataset Reduces Accuracy
**Risk:** n=28 more diverse than n=9, expect accuracy degradation (100% → 85-90%)
**Probability:** HIGH
**Impact:** ACCEPTABLE - This is expected and desirable for generalization
**Mitigation:** Set realistic success criteria (F1 ≥ 0.85 instead of 1.0)

### Risk 3: Cash Flow Data Quality Varies
**Risk:** Not all REITs provide detailed ACFO/AFCF breakdowns in financial statements
**Probability:** MEDIUM
**Mitigation:** Use Phase 1-3 pipeline to extract, validate against issuer-reported values
**Fallback:** Use subset of REITs with high-quality cash flow data (n=15-20 instead of n=28)

### Risk 4: Overfitting with 36 Features
**Risk:** 36 features on n=28 samples may cause overfitting
**Probability:** MEDIUM
**Mitigation:** Feature selection via SHAP importance, regularization, cross-validation
**Fallback:** Reduce to top 15-20 features based on importance ranking

---

## Related Issues

- Issue #37: Distribution Cut Prediction Model v1.0 (COMPLETE ✅ - 100% accuracy on n=9)
- Issue #39: OpenBB Platform Integration (COMPLETE ✅ - Automation enables v2.0)
- Issue #40: Phase 4 Template Integration (PENDING - Optional enhancement)

---

## Notes

**Why This Matters:**

The v1.0 model achieves perfect accuracy on n=9 by learning "high leverage + weak coverage = cut risk". This is correct but incomplete.

Version 2.0 will answer:
1. **"Why does high leverage matter?"** → Because debt service consumes free cash flow (AFCF)
2. **"When does leverage become dangerous?"** → When rates rise, macro weakens, or refinancing cliffs loom
3. **"How much runway does the REIT have?"** → Cash runway and burn rate analysis
4. **"What does the market think?"** → Price stress and volatility as early warning ✨ NEW
5. **"What's the management track record?"** → Recovery pattern analysis ✨ NEW

**This transforms the model from a pattern recognizer to an economic mechanism explainer.**

**Key Insight from Issue #39 Integration:**

Market risk signals and recovery patterns were NOT part of the original Issue #38 plan, but Issue #39 delivered these as unexpected bonuses. The hypothesis that market signals may provide early warning (1-2 quarters ahead of fundamentals) is worth testing and could significantly improve model performance.

---

**Created:** 2025-10-22
**Updated:** 2025-10-22 (Post Issue #39 Completion)
**Priority:** Medium (v1.0 is production-ready, v2.0 is enhancement)
**Original Effort:** 10 weeks
**Revised Effort:** 5-6 weeks (40-50% reduction via automation)
**Complexity:** MODERATE (automation reduces data collection burden significantly)
